# Tarea 1 

Alumnos:

* Miguel Espinoza
* Sebastián López
* Pedro Pérez

## P1

### Parte 1

In [1]:
import os

folder = 'data/raw/'  #Carpeta principal en la que estan los archivos
weeks = os.listdir(folder)  #Se agregan las subcarpetas a una lista
weeks

['w13', 'w14', 'w15', 'w16', 'w17']

In [2]:
#Se crean dos listas para guardar los archivos normales
#y los 'furnished'
files = []
files_furnished = []

#Mediante un bucle for se guardan las direcciones de todos
#los archivos en las listas creadas
for week in weeks:
    direccion = folder + week + '/'
    file = os.listdir(direccion)
    files.append(direccion + file[0])
    files_furnished.append(direccion + file[1])
    
files

['data/raw/w13/metrocuadrado_all_w13.csv',
 'data/raw/w14/metrocuadrado_all_w14.csv',
 'data/raw/w15/metrocuadrado_all_w15.csv',
 'data/raw/w16/metrocuadrado_all_w16.csv',
 'data/raw/w17/metrocuadrado_all_w17.csv']

In [3]:
import pandas as pd

#Se lee el primer dataframe para observar los datos
data_all = pd.read_csv(files[0])
data_all.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

En primer lugar se combinan los dataframes normales

In [4]:
data_all = pd.read_csv(files[0])

for file in files:
    dato_semanal = pd.read_csv(file)
    data_all = data_all.merge(dato_semanal, how = 'outer') 
    #Se pone 'outer' para hacer la union de todos los dataframes

print('Cantidad de datos: ', len(data_all))
data_all.head(3)

Cantidad de datos:  16470


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

In [5]:
#Se ve si hay duplicados
print('Duplicados: ', data_all.duplicated().sum())

Duplicados:  175


In [6]:
#Se eliminan las filas duplicadas
data_all.drop_duplicates(inplace=True)
print('Cantidad de datos: ', len(data_all))

Cantidad de datos:  16295


Ahora se combinan los archivos furnished

In [7]:
data_furnished = pd.read_csv(files_furnished[0])

for file in files_furnished:
    dato_semanal = pd.read_csv(file)
    data_furnished = data_furnished.merge(dato_semanal, how='outer')

print('Cantidad de datos: ', len(data_furnished))
data_furnished.head(3)

Cantidad de datos:  2101


property_type|rent_type|location       price  n_rooms  \
0  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000      3.0   
1  Casa en Venta Y Arriendo, BELLAVISTA Bogotá D.C..  $1.300.000      3.0   
2  Casa en Arriendo, PORTALES DEL NORTE Bogotá D.C..  $1.800.000      3.0   

   n_bath   surface                                            details  \
0     2.0  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
1     2.0   72.0 m2  hermosa casa en venta y en arriendo en, mandal...   
2     3.0  76.81 m2  Estrato 4. CASA TRIPLEX!! Exterior . Cuenta co...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
1  https://www.metrocuadrado.com/inmueble/venta-c...                 41.4  
2  https://www.metrocuadrado.com/inmueble/arriend...                 41.3

In [8]:
#Se buscan los duplicados y se eliminan
print('Duplicados: ', data_furnished.duplicated().sum())

Duplicados:  2


In [9]:
data_furnished.drop_duplicates(inplace=True)

print('Cantidad de datos: ', len(data_furnished))
#Se observa que son muchos menos los datos de tipo furnished

Cantidad de datos:  2099


Al intentar combinar ambos dataframes usando `data_all.merge(data_furnished, how='outer')` aparece el error:

`ValueError: You are trying to merge on object and float64 columns. If you wish to proceed you should use pd.concat`

Por lo que se decide comprobar los tipos de las columnas

In [10]:
#Se observan las columnas de los dataframe
print('Tipos Data All')
print(data_all.dtypes)
print()
print('Tipos Data Furnished')
print(data_furnished.dtypes)

Tipos Data All
property_type|rent_type|location     object
price                                object
n_rooms                              object
n_bath                               object
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object

Tipos Data Furnished
property_type|rent_type|location     object
price                                object
n_rooms                             float64
n_bath                              float64
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object


Se observa que el error se debe a que en `Data All` las columnas `n_rooms` y `n_bath` son de tipo `object`, en cambio en `Data Furnished` son de tipo `float64`.

Este problema se resuelve simplemente cambiando el tipo de esas columnas.

In [11]:
#Se cambian las columnas de 'data_furnished' a object
data_furnished = data_furnished.astype({'n_rooms':'object', 'n_bath':'object'})

#Obs: No se pueden cambiar las columnas de 'data_all' a float64, pues hay un dato llamado '5+'

In [153]:
#Se combinan los dataframes
data = data_all.merge(data_furnished, how='outer', indicator='Furnished')

#Se coloca utiliza el parámetro indicator para indicar de que dataframe 
#proviene cada fila.
#left_only indica que viene del dataframe 'data_all'
#right_only indica que viene del dataframe 'data_furnished'
#y both que proviene de ambos

print('Cantidad de datos: ', len(data))
data.head(3)

Cantidad de datos:  18394


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

   Furnished  
0  left_only  
1  left_only  
2  left_only

In [154]:
data['Furnished'].replace({'left_only':'No', 'right_only':'Si', 'both':'Ambos'}, inplace=True)

#Se cambian los nombres para indicar más claramente si provienen de furnished o no
data.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished  
0        No  
1        No  
2        No

In [155]:
#Se cuentan la cantidad de cada valores de cada tipo
data['Furnished'].value_counts()

No    16295
Si     2099
Name: Furnished, dtype: int64

Como se observa no hay elementos comunes entre ambos grupos, son disjuntos. No hay archivos con texto 'furnished' que no estén contenidos en archivos con texto 'all'.

In [156]:
#Esto se comprueba tambien haciendo una intersección entre los dataframes
#mediante el argumento 'inner'
pd.merge(data_all, data_furnished, how='inner')

#Se observa que el resultado es un dataframe vacío

Empty DataFrame
Columns: [property_type|rent_type|location, price, n_rooms, n_bath, surface, details, url, metrocuadrado_index]
Index: []

### Parte 2

#### A

Primero se limpia el precio

In [157]:
data['price'].head(3)

0    $1.050.000
1    $3.930.000
2      $750.000
Name: price, dtype: object

Se observa que son de tipo `object`, pues el precio contiene el signo peso '$' y los puntos. Convendría eliminar este signo y los puntos para poder pasar los datos a `float64`.

In [158]:
data['price'] = data['price'].str.replace('$','')
data['price'] = data['price'].str.replace('.','')
data['price'].head(3)

0    1050000
1    3930000
2     750000
Name: price, dtype: object

In [159]:
#Se cambia el tipo a float64
data['price'] = data['price'].astype('float')
data['price'].head(3)

0    1050000.0
1    3930000.0
2     750000.0
Name: price, dtype: float64

Ahora se limpia el área

In [160]:
data['surface'].tail(4)

18390    164.34 m2
18391     110.0 m2
18392     106.0 m2
18393     135.0 m2
Name: surface, dtype: object

Se deben eliminar los carácteres ' m2' al final del número, y luego cambiar el tipo a `float`.

In [161]:
data['surface'] = data['surface'].str.replace(' m2','')
data['surface'] = data['surface'].astype('float')
data['surface'].tail(4)

18390    164.34
18391    110.00
18392    106.00
18393    135.00
Name: surface, dtype: float64

Ahora se limpia el número de habitaciones

In [162]:
data['n_rooms'].head(3)

0    3
1    3
2    3
Name: n_rooms, dtype: object

Parece que fueran números enteros, pero el tipo de la columna es object

In [163]:
#Se ven los valores de la columna
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, '5+', 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

Se observa que la columna tiene valores de tipo string y float. Más aún se ve que se podría cambiar inmediatamente el tipo a `float` sino existiera el valor `5+`.

In [164]:
#Se cuenta cuantas entradas hay de cada valor
data['n_rooms'].value_counts()

2      6571
1      6271
3      1351
4      1198
1.0    1040
2.0     938
5       862
3.0      52
5.0      36
4.0      32
5+        9
Name: n_rooms, dtype: int64

Se ve que solo son 9 las entradas con valor `5+`, por lo que es una parte infima del total del dataset que es de $18394$ entradas, simplemente se podrían eliminar. Pero como no se desea botar datos, se cambiará el valor `5+` por $6$, dado que es el entero mayor a 5 más cercano.

In [165]:
data['n_rooms'] = data['n_rooms'].replace('5+',6)
#Se reemplazan el valor que daba problemas, y se muestran los valores de las entradas
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, 6, 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

In [166]:
#Como esta correcto, se puede cambiar el tipo de la columna
data['n_rooms'] = data['n_rooms'].astype('float')
data['n_rooms'].head(3)

0    3.0
1    3.0
2    3.0
Name: n_rooms, dtype: float64

Por último se limpia el número de baños

In [167]:
data['n_bath'].tail(3)

18391    3
18392    3
18393    2
Name: n_bath, dtype: object

Nuevamente parecen ser enteros, pero el tipo es object

In [168]:
#Se ven los valores de la columna
data['n_bath'].unique()

array(['2', '1', '4', '3', nan, '5', '5+', 2.0, 3.0, 4.0, 5.0, 1.0],
      dtype=object)

Nuevamente el error viene dado por `5+`, por lo que se procede de igual manera que en `n_rooms`.

In [169]:
data['n_bath'] = data['n_bath'].replace('5+',6) #Se reemplazan '5+' por 6
data['n_bath'] = data['n_bath'].astype('float') #Se cambia el tipo a float
data['n_bath'].head(3)

0    2.0
1    2.0
2    1.0
Name: n_bath, dtype: float64

Finalmente se muestran nuevamente los tipos de las columnas

In [170]:
data.dtypes

property_type|rent_type|location     object
price                               float64
n_rooms                             float64
n_bath                              float64
surface                             float64
details                              object
url                                  object
metrocuadrado_index                 float64
Furnished                            object
dtype: object

#### B

Se debe dividir la columna `property_type|rent_type|location` en 3 columnas distintas.

In [171]:
#Se hace una copia de la data por si se realiza un error
copia_seguridad = data.copy()

In [151]:
#Cuando algo salga mal correr esta celda
data = copia_seguridad.copy()

In [172]:
#Se muestran unos pocos datos de esta columna
data['property_type|rent_type|location']

0          Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..
1        Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...
2        Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...
3              Casa en Arriendo, PARIS GAITAN Bogotá D.C..
4        Casa en Arriendo, BALCONES DE ORIENTE Bogotá D...
                               ...                        
18389    Apartamento en Arriendo, CHICO NORTE Bogotá D.C..
18390    Apartamento en Arriendo, CHICO NORTE II Bogotá...
18391    Apartamento en Venta Y Arriendo, CHICO NORTE B...
18392    Apartamento en Arriendo, MOLINOS NORTE Bogotá ...
18393       Apartamento en Arriendo, EL NOGAL Bogotá D.C..
Name: property_type|rent_type|location, Length: 18394, dtype: object

Se pueden deducir varias cosas:
1. El tipo de propiedad es la primera palabra del string.
2. El tipo de renta aparece después de la palabra 'en' y un espacio ' '.
3. El lugar esta después de la coma.

En primer lugar se decide dividir el string en dos partes separadas por la coma.

In [173]:
new_columns = data['property_type|rent_type|location'].str.split(pat=',', expand=True)
new_columns.head(3)

0                                           1
0  Casa en Arriendo               LA SOLEDAD NORTE Bogotá D.C..
1  Casa en Arriendo    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..
2  Casa en Arriendo   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

In [174]:
#Se insertan las columnas en la data 
data[['property_type|rent_type', 'location']] = new_columns

#Se elimina la columna original
data.drop('property_type|rent_type|location', axis=1, inplace=True)

data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished property_type|rent_type  \
0        No        Casa en Arriendo   
1        No        Casa en Arriendo   
2        No        Casa en Arriendo   

                                     location  
0               LA SOLEDAD NORTE Bogotá D.C..  
1    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..  
2   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

Ahora se divide la columna `property_type|rent_type` en dos columnas más.

In [175]:
new_columns = data['property_type|rent_type'].str.split(pat='en ', expand=True)
new_columns.head(3)

0         1
0  Casa   Arriendo
1  Casa   Arriendo
2  Casa   Arriendo

In [176]:
#Se insertan las columnas en la data y se elimina la columna original
data[['property','rent']] = new_columns
data.drop('property_type|rent_type', axis=1, inplace=True)
data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished                                    location property      rent  
0        No               LA SOLEDAD NORTE Bogotá D.C..    Casa   Arriendo  
1        No    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..    Casa   Arriendo  
2        No   VILLA DE LOS ALPES II SECTOR Bogotá D.C..    Casa   Arriendo

Por último se revisa que no hayan espacios vacíos al inicio o final de las nuevas columnas.

In [178]:
print(data['location'][0])
print(data['property'][0])
print(data['rent'][0])

 LA SOLEDAD NORTE Bogotá D.C..
Casa 
Arriendo


Tanto location como property tienen espacios, por lo que se utiliza strip para eliminarlos.

In [182]:
data['location'] = data['location'].str.strip()
data['property'] = data['property'].str.strip()

### Parte 3

### Parte 4

### Parte 5

### Parte 6

## P2

# P3 

## Implementación

### Parte 1

In [79]:
from sklearn.base import BaseEstimator as BE
from sklearn.base import RegressorMixin as RM
import numpy as np

class RegresionBayesianaEmpirica(BE,RM):
    
    def __init__(self,alpha_0,beta_0,tol=10**(-5),maxiter=200):
        self.__alpha = alpha_0
        self.__beta = beta_0
        self.__tol = tol
        self.__maxiter = maxiter
        
    def get_posteriori(self,X,y,alpha,beta):
        N,d = X.shape
        S_N = np.linalg.inv(alpha*np.eye(d) + beta*np.dot(X.T,X)) # dim dxd
        m_N = beta*np.dot(S_N,np.dot(X.T,y)) # dim dx1
        #K = np.exp( (-1/2)*(beta*np.dot(y.T,y) - (beta**2)*np.dot(y.T,np.dot(X,np.dot(S_N,np.dot(X.T,y))))) )*(np.linalg.det(S_N)**0.5)*((2*np.pi)**(d/2 - N))*(alpha**(d/2))*(beta**(N/2))
        
        #def dens_posteriori(w):
            # w de dim dx1
            #return K*((2*np.pi)**(-d/2))*(np.linalg.det(S_N)**(-0.5))*np.exp(-0.5*(((w-m_N).T)*(np.linalg.inv(S_N))*(w-m_N)))
        
        return m_N, S_N#, dens_posteriori
    
    
    def fit(self,X,y,i=0):
        i += 1
        print('Iteracion',i)
        print('Valores actuales: alpha=',self.__alpha,' beta=',self.__beta)
        if i <= self.__maxiter:
            N,d = X.shape
            eigen_val, eigen_vec = np.linalg.eig(self.__beta*np.dot(X.T,X))
            gamma = np.sum(eigen_val/(self.__alpha + eigen_val))
            m_N, S_N = self.get_posteriori(X,y,self.__alpha,self.__beta)
        
            new_alpha = gamma/(np.dot(m_N.T,m_N))
            new_beta = (N-gamma)/(np.sum((y-np.dot(X,m_N))**2))
        
            if np.abs(self.__alpha - new_alpha) < self.__tol and np.abs(self.__beta - new_beta) < self.__tol:
                self.__alpha = new_alpha
                self.__beta = new_beta
                self.__m_N = m_N
                self.__S_N = S_N
            else:
                self.__alpha = new_alpha
                self.__beta = new_beta
                self.__m_N = m_N
                self.__S_N = S_N
                self.fit(X,y,i)
                
    
    def predict(self,X_,return_std=False):
        y_ = np.dot(X_,self.__m_N)
        y_std = np.sqrt((1/self.__beta)*np.dot(X_,np.dot(self.__S_N,X_.T)))
        
        if return_std:
            return (y_,y_std)
        return y_

### Parte 2

In [82]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder, PolynomialFeatures, FunctionTransformer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

def data_1d(df_column):
    return df_column.values.reshape([-1,1])

#reshape
rs = FunctionTransformer(data_1d)

#cat pipe
cat_pipe = Pipeline([('reshape', rs),('ordinal', OneHotEncoder(handle_unknown='ignore'))])

#num pipe
num_pipe = Pipeline([('reshape', rs),('scaler',StandardScaler()),('poly',PolynomialFeatures(degree=3))])

#Se diferencian columnas numéricas y categóricas
prep = ColumnTransformer(transformers=[('num',num_pipe,num_cols),('cat',cat_pipe,cat_cols)])

NameError: name 'num_cols' is not defined

### Parte 3

In [84]:
from sklearn.model_selection import train_test_split

def run(df_X, df_y, pipeline):
    X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2)
    alpha_ini = 1e-7
    beta_ini = 1e-8
    
    pipeline.fit(X_train,y_train)
    y_pred = pipeline.predict(X_test) 
    
    #Tubo = Pipeline([('prep',prep),('clf', RegressionBayesianaEmpirica(alpha_ini,beta_ini))])
    #Tubo.fit(X_train, y_train)
    #y_pred = Tubo.predict(X_test)
    
    R_2 = Tubo.score(X_test, y_test)
    RMSE = np.sqrt(np.mean((y_pred-y_test)**2))
    print('RMSE:',RMSE,' R_2:',R_2)

NameError: name 'df_X' is not defined

### Parte 4

In [85]:
Tubo = Pipeline([('prep',prep),('clf', RegressionBayesianaEmpirica(alpha_ini,beta_ini))])
run(df_X,df_y,Tubo)

NameError: name 'prep' is not defined

### Parte 5

In [86]:
from sklearn.linear_model import BayesianRidge

BR_pipe = Pipeline([('prep',prep),('clf',BayesianRidge())])
run(df_x,df_y,BR_pipe)

NameError: name 'prep' is not defined